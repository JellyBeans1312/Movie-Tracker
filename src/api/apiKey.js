export const apiKey = '044b86cc1259591cb5a872bda8d25';