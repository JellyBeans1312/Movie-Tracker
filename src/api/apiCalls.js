  // const url = 'http://localhost:3000'
import { apiKey } from './apiKey'

export const fetchMovies = () => {
  return fetch(`https://api.themoviedb.org/3/movie/now_playing?api_key=${apiKey}edd&language=en-US&page=1`)
    .then(res => res.json())
    .then(res => res.results)
    .catch (error => error.message)   
}

export const addUser = async (user) => {
  try {
    const options = {
      method: 'POST',
      body: JSON.stringify(user),
      headers: {
        'Content-Type': 'application/json'
      }
    }

    const response = await fetch('http://localhost:3000/api/users/new', options)
  
    return response.ok;
  } catch (error) {
    console.log(error.message);
    throw new Error("Failed to post error");
  }
}

export const getUser = async (user) => {
  try {
    const options = {
      method: 'POST',
      body: JSON.stringify(user),
      headers: {
        'Content-Type': 'application/json'
      }
    }
   
    const response = await fetch('http://localhost:3000/api/users', options)
    const result = await response.json()
    
    return result.data
  } catch (error) {
    throw new Error("Email and password do not match");
  } 
}

export const favoriteMovie = async (movieInfo) => {
  try {
    const options = {
      method: 'POST',
      body: JSON.stringify(movieInfo),
      headers: {
        'Content-Type': 'application/json'
      }
    }

    await fetch(`http://localhost:3000/api/users/favorites/new`, options)
  
  } catch (error) {
    throw new Error("Failed to fetch favorites");
  }
}

export const fetchFavorites = async (userId) => {
  try {
    const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites`);
    const result = await response.json();
    return result.data
  } catch (error) {
    throw new Error('Failed to fetch favorites');
  }
}

export const removeFavorite = async (userId, movieId) => {
  try {
    const option = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    }

    await fetch(`http://localhost:3000/api/users/${userId}/favorites/${movieId}`, option);
    
  } catch (error) {
    throw new Error('Failed to remove favorite')
  }
}