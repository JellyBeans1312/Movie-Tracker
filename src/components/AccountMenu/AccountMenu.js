import React, { Component } from 'react';
import { Login } from '../Login/Login';
import { login, showError } from '../../actions';
import { connect } from 'react-redux';
import { Redirect } from 'react-router'
import { getUser } from '../../api/apiCalls'
import { fetchFavorites } from '../../api/apiCalls';
import { setFavorites } from '../../actions'
import './AccountMenu.css'
import PropTypes from 'prop-types';

export class AccountMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: '',
      password: '',
      display: 'start',
      error: null
    };
  }
  
  handleChange = (e) => {
    const {name, value} = e.target
    this.setState({ [name]: value })
  }

  handleSubmit = async (e) => {
    try {
      e.preventDefault()
      await getUser(this.state)
      .then(data => this.props.login(data))
      this.setState({display: 'loggedIn'})
      await fetchFavorites(this.props.user.id)
      .then(result => this.props.setFavorites(result))
    } catch (error) {
      this.setState({ error: error.message })
    }
  }

  render() {
    const isLoggedIn = this.state.display === 'loggedIn'
    let view;
    
    if(!isLoggedIn) {
      view = <Login email={this.state.email} password={this.state.password} handleChange={this.handleChange} handleSubmit={this.handleSubmit}/>
    } else {
      view = <Redirect to='/LoggedIn'/>
    } 
    return (  
      <div className='log-in-bar'>
        {view}
        {this.state.error && <h2 className='error'>{this.state.error}</h2>}
      </div>
    )
  }
};

export const mapStateToProps = (store) => ({
  user: store.login
});

export const mapDispatchToProps = (dispatch) => ({
  login: (user) => dispatch(login(user)),
  setFavorites: (favorites) => dispatch(setFavorites(favorites))
});

AccountMenu.propTypes = {
  user: PropTypes.object.isRequired,
  login: PropTypes.func.isRequired,
  favorites: PropTypes.array.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(AccountMenu)
