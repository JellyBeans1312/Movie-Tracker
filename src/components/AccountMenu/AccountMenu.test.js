import { shallow } from 'enzyme';
import { AccountMenu } from './AccountMenu';
import { getUser } from '../../api/apiCalls'
import React from 'react';
import { login, setFavorites } from '../../actions';
import { mapStateToProps, mapDispatchToProps } from './AccountMenu';

jest.mock('../../api/apiCalls', () => ({
  getUser: jest.fn().mockImplementation(() => {
    return { id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com" }
  })
}));

describe('accountMenu', () => {
  
  describe('accountMenu component', () => {
    let wrapper;
    let instance;
    let user;

    beforeEach(() => {
      user = { id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com" };
      wrapper = shallow(<AccountMenu user={user} />);
      instance = wrapper.instance();
    })
  
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it.skip('should change state when handleChange is invoked', () => {
      const e = {
        target: {
          name: 'email',
          value: 'blah@blah.com'
        }
      }
      instance.handleChange(e)
      expect(wrapper.state('email')).toEqual('blah@blah.com')
    });
  
    it.skip('should change state of display when handleSubmit resolves', async () => {
      const e = {
        preventDefault: jest.fn()
      }
      expect(wrapper.state('display')).toEqual('start');
      await instance.handleSubmit(e);
      expect(wrapper.state('display')).toEqual('loggedIn');
    });
  
    it.skip('should set state of error when handleSubmit catches an error', () => {
      const e = {
        preventDefault: jest.fn()
      }
      window.fetch = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          ok: false
        })
      })
      instance.handleSubmit(e);
      expect(wrapper.state('error')).toEqual("Email and password do not match");
    });
  })


  describe('mapStateToProps', () => {

    it('should return an object with user info', () => {
      const mockStore = {
        login: { id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com" },
      }
  
      const expected = {
        user: { id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com" },
      }

      const mappedProps = mapStateToProps(mockStore);

      expect(mappedProps).toEqual(expected)
    });
  });

  describe('dispatchStateToProps', () => {
    
    it('calls dispatch with login action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setFavorites([{title: 'spider-man'}, {name: 'lion king'}])

      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.setFavorites([{ title: 'spider-man' }, { name: 'lion king' }])

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with setFavorites action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = login({ name: "Taylor", password: "password", email: "tman2272@aol.com" })

      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.login({ name: "Taylor", password: "password", email: "tman2272@aol.com" })

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  });

})