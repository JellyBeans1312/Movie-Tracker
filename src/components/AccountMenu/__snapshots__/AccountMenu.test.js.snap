// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountMenu accountMenu component should match the snapshot 1`] = `
<div
  className="log-in-bar"
>
  <Login
    email=""
    handleChange={[Function]}
    handleSubmit={[Function]}
    password=""
  />
</div>
`;
